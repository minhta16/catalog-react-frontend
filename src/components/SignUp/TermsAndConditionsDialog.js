import React from 'react';
import PropTypes from 'prop-types';

import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Button,
  DialogContentText,
} from '@material-ui/core';

/**
 * This component renders a mock terms and conditions dialog
 */
export function TermsAndConditionsDialog(props) {
  const { open, onClose } = props;
  const terms = `Cras mattis consectetur purus sit amet fermentum.
  Cras justo odio, dapibus ac facilisis in, egestas eget quam.
  Morbi leo risus, porta ac consectetur ac, vestibulum at eros.
  Praesent commodo cursus magna, vel scelerisque nisl consectetur et.`;
  return (
    <Dialog open={open} onClose={onClose}>
      <DialogTitle>Terms and Conditions</DialogTitle>
      <DialogContent dividers>
        <DialogContentText>{terms}</DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>Got it!</Button>
      </DialogActions>
    </Dialog>
  );
}

TermsAndConditionsDialog.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
};

TermsAndConditionsDialog.defaultProps = {
  onClose: () => {},
  open: true,
};

export default TermsAndConditionsDialog;
