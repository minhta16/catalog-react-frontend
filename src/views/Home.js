import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { Container } from '@material-ui/core';
import CategoriesTable from 'components/CategoriesTable';
import 'App.css';

import { fetchCategories as fetchCategoriesRedux } from 'references/redux/actions/categories';
import { fetchPosts as fetchPostsRedux } from 'references/redux/actions/posts';
import { selectAllPosts } from 'references/redux/selectors/posts';
import { selectCategories, selectCategory } from 'references/redux/selectors/categories';

export class Home extends Component {
  componentDidMount() {
    const { fetchCategories, match, fetchPosts } = this.props;
    fetchCategories();
    if (match.params.id) {
      fetchPosts(match.params.id);
    }
  }

  render() {
    const { match, categories, selectedCatItems, fetchPosts, selectedCategory } = this.props;
    return (
      <Container maxWidth="lg">
        <CategoriesTable
          categories={categories}
          selectedCatId={match.params.id}
          selectedCat={selectedCategory}
          selectedCatItems={selectedCatItems}
          fetchPosts={fetchPosts}
        />
      </Container>
    );
  }
}

Home.propTypes = {
  match: PropTypes.object,
  categories: PropTypes.array,
  fetchCategories: PropTypes.func.isRequired,
  fetchPosts: PropTypes.func.isRequired,
  selectedCatItems: PropTypes.array,
  selectedCategory: PropTypes.object,
};

Home.defaultProps = {
  match: {},
  categories: {},
  selectedCatItems: {},
  selectedCategory: {},
};

const mapSelectorToProps = (state, ownProps) => ({
  selectedCatItems: selectAllPosts(state),
  categories: selectCategories(state),
  selectedCategory: selectCategory(state, ownProps.match.params.id),
});

const mapDispatchToProps = {
  fetchCategories: fetchCategoriesRedux,
  fetchPosts: fetchPostsRedux,
};

export default connect(
  mapSelectorToProps,
  mapDispatchToProps,
)(Home);
