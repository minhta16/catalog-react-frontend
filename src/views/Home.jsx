import React, { Component } from 'react';
import { connect } from 'react-redux';

import { Container } from '@material-ui/core';
import CategoriesTable from 'components/CategoriesTable';
import 'App.css';

class Home extends Component {
  componentDidMount() {
    return 0;
  }

  render() {
    return (
      <Container maxWidth="lg">
        <CategoriesTable />
      </Container>
    );
  }
}

const mapStateToProps = state => ({
  categories: state.categories,
});

export default connect(mapStateToProps)(Home);
