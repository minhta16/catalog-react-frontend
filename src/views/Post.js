import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import PostPaper from 'components/PostPaper';
import { Container } from '@material-ui/core';

export function Post(props) {
  const { match, selectedCatItems } = props;
  const currentPost = selectedCatItems[match.params.postId];
  return (
    <Container maxWidth="lg">
      <PostPaper
        header={currentPost.name}
        created={currentPost.created}
        body={currentPost.description}
        categoryId={match.params.id}
      />
    </Container>
  );
}

Post.propTypes = {
  match: PropTypes.shape.isRequired,
  selectedCatItems: PropTypes.shape.isRequired,
};

const mapStateToProps = (state) => ({
  selectedCatItems: state.posts,
});
Post.propTypes = {};

export default connect(mapStateToProps)(Post);
