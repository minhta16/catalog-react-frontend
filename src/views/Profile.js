import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import UserPostsPaper from 'components/UserPostsPaper';
import { Container } from '@material-ui/core';
import { selectCurrentUser } from 'references/redux/selectors/users';

export function Post(props) {
  const { currentUser } = props;
  return (
    <Container maxWidth="lg">
      <UserPostsPaper username={currentUser.username} token={currentUser.token} />
    </Container>
  );
}

Post.propTypes = {
  currentUser: PropTypes.shape.isRequired,
};

const mapStateToProps = (state) => ({
  currentUser: selectCurrentUser(state),
});
Post.propTypes = {};

export default connect(mapStateToProps)(Post);
