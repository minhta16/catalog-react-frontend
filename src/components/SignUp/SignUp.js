/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { TextField, Paper, Container, Typography, Button, Link } from '@material-ui/core';
import TermsAndConditionsDialog from 'components/SignUp/TermsAndConditionsDialog';
import { createUserAndSignIn as createUserAndSignInRedux } from 'actions/users';

export class SignUp extends Component {
  state = {
    username: '',
    name: '',
    email: '',
    password: '',
    confirmPassword: '',
    emailWarning: true,
    passWarning: true,
    passwordMatchWarning: true,
    openTerms: false,
    redirect: false,
  };

  /**
   * Connect components to state
   */
  handleChange = (e) => {
    this.setState(
      {
        [e.target.id]: e.target.value,
      },
      this.handleCreateWarnings,
    );
  };

  /**
   * Opens terms and conditions upon called
   */
  handleTermsClick = () => {
    const { openTerms } = this.state;
    this.setState({
      openTerms: !openTerms,
    });
  };

  /**
   * create an user and sign in
   */
  register = () => {
    const { username, password, email, name, passwordMatchWarning, passWarning } = this.state;
    const { createUserAndSignIn } = this.props;
    if (!passwordMatchWarning && !passWarning) {
      createUserAndSignIn(username, password, email, name);
      this.setState({
        redirect: true,
      });
    }
    // handle error here
  };

  /**
   * returns true if an email is qualified
   */
  qualifiedEmail = (email) => {
    return /^[a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9]\.[a-zA-Z]{2,}$/.test(email);
  };

  /**
   * returns true if password is 0 character long (user hasn't enter a password yet) or if the password is qualified (contains a number, character, and more than 8 chars long)
   */
  qualifiedPassword = (password) => {
    return (
      password.length === 0 ||
      (password.length >= 8 && /\d/.test(password) && /[A-z]/.test(password))
    );
  };

  /**
   * Create warnings if user has unqualified things in text fields
   */
  handleCreateWarnings() {
    const { email, password, confirmPassword } = this.state;

    this.setState({
      emailWarning: !this.qualifiedEmail(email),
    });
    this.setState({
      passWarning: !this.qualifiedPassword(password),
    });
    this.setState({
      passwordMatchWarning: password !== confirmPassword,
    });
  }

  render() {
    const {
      username,
      name,
      email,
      password,
      confirmPassword,
      openTerms,
      emailWarning,
      passWarning,
      passwordMatchWarning,
      redirect,
    } = this.state;
    const termsText = 'By clicking Resgister, you agree with our ';
    if (redirect) {
      return <Redirect exact to="/" />;
    }
    return (
      <Container maxWidth="lg">
        <Paper className="paper">
          <Typography variant="h4">Register now!</Typography>
          <form autoComplete="off">
            <TextField
              required
              id="username"
              label="Username"
              value={username}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              id="name"
              label="Name"
              value={name}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              error={emailWarning}
              helperText={emailWarning ? `Email needs to match the format example@domain.com` : ''}
              id="email"
              label="Email Address"
              value={email}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              error={passWarning}
              helperText={
                passWarning
                  ? `Password needs to have at least 8 characters AND at least 1 number`
                  : ''
              }
              id="password"
              label="Password"
              value={password}
              onChange={this.handleChange}
              type="password"
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              error={passwordMatchWarning}
              helperText={passwordMatchWarning ? `Password doesn't match` : ''}
              id="confirmPassword"
              label="Confirm Password"
              value={confirmPassword}
              onChange={this.handleChange}
              type="password"
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <Typography variant="subtitle2">
              {termsText}
              <Link id="sign-up-terms-and-condition" onClick={this.handleTermsClick}>
                Terms and Conditions
              </Link>
            </Typography>
            <Button
              variant="contained"
              color="primary"
              onClick={this.register}
              disabled={passWarning || passwordMatchWarning}
            >
              Register
            </Button>
          </form>
          <TermsAndConditionsDialog open={openTerms} onClose={this.handleTermsClick} />
        </Paper>
      </Container>
    );
  }
}

SignUp.propTypes = {
  createUserAndSignIn: PropTypes.func.isRequired,
};
const mapDispatchToProps = {
  createUserAndSignIn: createUserAndSignInRedux,
};

export default connect(
  null,
  mapDispatchToProps,
)(SignUp);
