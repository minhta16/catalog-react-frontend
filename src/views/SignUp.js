/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { TextField, Paper, Container, Typography, Button, Link } from '@material-ui/core';
import TermsAndConditionsDialog from 'components/TermsAndConditionsDialog';
import { createUserAndSignIn as createUserAndSignInRedux } from 'references/redux/actions/users';

export class SignUp extends Component {
  state = {
    username: '',
    name: '',
    email: '',
    password: '',
    confirmPassword: '',
    emailWarning: true,
    passWarning: true,
    passwordMatchWarning: true,
    openTerms: false,
    redirect: false,
  };

  handleChange = (e) => {
    this.setState(
      {
        [e.target.id]: e.target.value,
      },
      this.handleCreateWarnings,
    );
  };

  handleTermsClick = () => {
    const { openTerms } = this.state;
    this.setState({
      openTerms: !openTerms,
    });
  };

  register = () => {
    const { username, password, email, name, passwordMatchWarning, passWarning } = this.state;
    const { createUserAndSignIn } = this.props;
    if (!(passwordMatchWarning || passWarning)) {
      createUserAndSignIn(username, password, email, name);
    }
    this.setState({
      redirect: true,
    });
    // handle error here
  };

  qualifiedEmail = (email) => {
    return /\w+@\w+\.\w+/.test(email);
  };

  qualifiedPassword = (password) => {
    return password.length >= 8 && /\d/.test(password) && /\w/.test(password);
  };

  handleCreateWarnings() {
    const { email, password, confirmPassword } = this.state;

    this.setState({
      emailWarning: !this.qualifiedEmail(email),
    });
    this.setState({
      passWarning: !this.qualifiedPassword(password),
    });
    this.setState({
      passwordMatchWarning: password !== confirmPassword,
    });
  }

  render() {
    const {
      username,
      name,
      email,
      password,
      confirmPassword,
      openTerms,
      emailWarning,
      passWarning,
      passwordMatchWarning,
      redirect,
    } = this.state;
    const termsText = 'By clicking Resgister, you agree with our ';
    if (redirect) {
      return <Redirect exact to="/" />;
    }
    return (
      <Container maxWidth="lg">
        <Paper className="paper">
          <Typography variant="h4">Register now!</Typography>
          <form autoComplete="off">
            <TextField
              required
              id="username"
              label="Username"
              value={username}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              id="name"
              label="Name"
              value={name}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              error={emailWarning}
              helperText={emailWarning ? `Email needs to match the format example@domain.com` : ''}
              id="email"
              label="Email Address"
              value={email}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              error={passWarning}
              helperText={
                passWarning
                  ? `Password needs to have at least 8 characters AND at least 1 number`
                  : ''
              }
              id="password"
              label="Password"
              value={password}
              onChange={this.handleChange}
              type="password"
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              error={passwordMatchWarning}
              helperText={passwordMatchWarning ? `Password doesn't match` : ''}
              id="confirmPassword"
              label="Confirm Password"
              value={confirmPassword}
              onChange={this.handleChange}
              type="password"
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <Typography variant="subtitle2">
              {termsText}
              <Link id="sign-up-terms-and-condition" onClick={this.handleTermsClick}>
                Terms and Conditions
              </Link>
            </Typography>
            <Button
              variant="contained"
              color="primary"
              onClick={this.register}
              disabled={passWarning || passwordMatchWarning}
            >
              Register
            </Button>
          </form>
          <TermsAndConditionsDialog open={openTerms} onClose={this.handleTermsClick} />
        </Paper>
      </Container>
    );
  }
}

SignUp.propTypes = {
  createUserAndSignIn: PropTypes.func.isRequired,
};
const mapDispatchToProps = {
  createUserAndSignIn: createUserAndSignInRedux,
};

export default connect(
  null,
  mapDispatchToProps,
)(SignUp);
