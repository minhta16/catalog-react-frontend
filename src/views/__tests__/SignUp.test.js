/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import { SignUp } from 'views/SignUp';

describe('views/SignUp', () => {
  let wrapper;
  let props;

  const setup = () => {
    wrapper = shallow(<SignUp {...props} />);
  };

  const update = () => {
    wrapper.update();
    wrapper.instance().forceUpdate();
  };

  beforeEach(() => {
    props = {
      createUserAndSignIn: jest.fn(),
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleTermsClick when click Terms and Condition', () => {
    setup();
    jest.spyOn(wrapper.instance(), 'handleTermsClick');
    update();
    const link = wrapper.find('#sign-up-terms-and-condition');
    link.simulate('click');
    expect(wrapper.instance().handleTermsClick).toHaveBeenCalled();
  });

  it('should call handleChange when a textfield is changed', () => {
    setup();
    jest.spyOn(wrapper.instance(), 'handleChange');
    update();
    const textField = wrapper.find('#username');
    textField.simulate('change', {
      target: {
        id: 'username',
        value: 'lol',
      },
    });
    expect(wrapper.instance().handleChange).toHaveBeenCalled();
  });
});
