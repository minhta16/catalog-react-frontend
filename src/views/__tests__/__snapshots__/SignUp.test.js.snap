// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/SignUp should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  maxWidth="lg"
>
  <WithStyles(ForwardRef(Paper))
    className="paper"
  >
    <WithStyles(ForwardRef(Typography))
      variant="h4"
    >
      Register now!
    </WithStyles(ForwardRef(Typography))>
    <form
      autoComplete="off"
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="username"
        label="Username"
        margin="normal"
        onChange={[Function]}
        required={true}
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        id="name"
        label="Name"
        margin="normal"
        onChange={[Function]}
        required={true}
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        error={true}
        fullWidth={true}
        helperText="Email needs to match the format example@domain.com"
        id="email"
        label="Email Address"
        margin="normal"
        onChange={[Function]}
        required={true}
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        error={true}
        fullWidth={true}
        helperText="Password needs to have at least 8 characters AND at least 1 number"
        id="password"
        label="Password"
        margin="normal"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        error={true}
        fullWidth={true}
        helperText="Password doesn't match"
        id="confirmPassword"
        label="Confirm Password"
        margin="normal"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(Typography))
        variant="subtitle2"
      >
        By clicking Resgister, you agree with our 
        <WithStyles(ForwardRef(Link))
          id="sign-up-terms-and-condition"
          onClick={[Function]}
        >
          Terms and Conditions
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        disabled={true}
        onClick={[Function]}
        variant="contained"
      >
        Register
      </WithStyles(ForwardRef(Button))>
    </form>
    <TermsAndConditionsDialog
      onClose={[Function]}
      open={false}
    />
  </WithStyles(ForwardRef(Paper))>
</WithStyles(ForwardRef(Container))>
`;
